### ISSUE TITLE
Cannot use `pub` with --preview-dart-2

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#32188]
```
> DART_VM_OPTIONS=--preview-dart-2 pub 
Invalid isolate state - Unable to make it runnable
```

On Dart VM version: 2.0.0-dev.26.0 (Thu Feb 15 03:06:13 2018 +0100) on "macos_x64"

**Comment by kevmoo** [comment]
CC @a-siva @nex3 @zanderso

**Comment by a-siva** [comment]
in the dart sdk builds pub, analyzer, dart2js and ddc use an application snapshot to run, these snapshots are built in dart 1 mode. When preview-dart-2 is specified it is not compatible with these snapshots and hence the failure. The error message is not very helpful and we will fix that.

To get these command with --preview-dart-2 first the pub, dart2js, analyzer code needs to be made strong mode clean and these application snapshots need to be generated with that option turned on. I think  we planned to do this in Q2.

The issue that the original bug wanted addressed was a way for folks to run tests in preview-dart-2 mode using pub. For this to work 'pub run' needs to implement the --preview-dart-2 option so that it starts the test process using this option, somewhat similar to the '-c' or '--checked' option that it currently supports.

sivalinuxmach[sdk]>out/ReleaseX64/dart-sdk/bin/pub run --help
Run an executable from a package.

Usage: pub run <executable> [args...]
-h, --help            Print this usage information.
-c, --[no-]checked    Enable runtime type checks and assertions.
    --mode            Mode to run transformers in.
                      (defaults to "release" for dependencies, "debug" for entrypoint)

**Comment by a-siva** [comment]
Who needs to own this issue of implementing the option '--preview-dart-2' in 'pub run' ?
Not sure who owns pub these days.

**Comment by mit-mit** [comment]
We have no plans, or bandwidth, to work on pub client in q1.

**Comment by zoechi** [comment]
related to dart-lang/pub#1807

**Comment by dgrove** [comment]
@nex3 @grouma @natebosch Can you all figure out what's needed here?

**Comment by jmesserly** [comment]
>/cc @kmillikin - there are likely changes need to CFE to make it run on Dart 2.

FYI, I was able to locally run DDC and DDK with `--preview-dart-2`. All tests were passing. Also DDC/DDK are now building themselves as part of their snapshot process (selfhost for training data), and that compile succeeds with no errors. Based on that I believe Analyzer/CFE are pretty Dart 2 clean at this point. (There might be some runtime cast failures on code paths that DDC/DDK does not exercise).

> To get these command with --preview-dart-2 first the pub, dart2js, analyzer code needs to be made strong mode clean and these application snapshots need to be generated with that option turned on. I think we planned to do this in Q2.

I haven't figured out how to build the DDC snapshots yet with `--preview-dart-2`. Would appreciate help with that. I've been chatting a bit with @a-siva

**Comment by nex3** [comment]
Once https://github.com/dart-lang/pub/pull/1871 lands, we won't need dart2js or DDC to be Dart 2-compatible anymore. We do still use the analyzer, though—is that Dart 2-clean?

**Comment by natebosch** [comment]
I'll work on checking that `pub run` code paths work in Dart 2 runtime.

I'll likely need some help getting it set up in the `pub` wrapping script in the SDK.

**Comment by nex3** [comment]
Is `test` Dart 2 clean? If so, we should be able to verify pub's cleanliness by running its tests.

If not, we can get part of the way by spawning the `pub` subprocesses the tests run with `--preview-dart-2`.

**Comment by natebosch** [comment]
`pub run` seems to work when I fix the static errors.
https://github.com/dart-lang/pub/pull/1874

However it's _drastically_ slower.

Set up a package which has a single dependency:

```yaml
dependencies:
  pub:
    path: ../pub
```

Invoke pub with no arguments - just prints usage and exits.

- `pub run pub` - 1.36 s
- `dart ../pub/bin/pub.dart` - 0.87s
- `dart --preview-dart-2 ../pub/bin/pub.dart` - 11.46s
- `dart --preview-dart-2 ../pub/bin/pub.dart run pub` - 18.35s

It's possible that using a snapshot in Dart 2 semantics will be an improvement, and that this will carry over when using a hosted dependency rather than a path dependency...

**Comment by natebosch** [comment]
I can't get timing data with snapshots - https://github.com/dart-lang/sdk/issues/32911

**Comment by natebosch** [comment]
Here is the list of work that I think is blocking the end to end `pub run --preview-dart-2 test` scenario.

- [ ] Sync pub into the SDK. Blocked on https://github.com/dart-lang/sdk/issues/32900
- [ ] Set up the flags in the shell and bat script to get the `--preview-dart-2` flag passed to the `dart` vm and not pub itself. (could skip this if we're ok with `DART_VM_OPTIONS=--preview-dart-2` ux)
- [ ] Ensure that `package:test` is Dart 2 runtime clean
- [ ] Give `pub` an understanding of whether the `run` command is Dart 2 or not. Today it wants to run from snapshots when it can, and Dart1/Dart2 snapshots are incompatible (and today not possible for Dart2, blocked on #32911). The hacky thing to do is generate only 1 snapshot and run from source when using the "other" mode, the full solution would involve generating2 snapshots and picking the right one at runtime.

**Comment by nex3** [comment]
Given that the running from source is so slow with the CFE right now, how do we verify that a given chunk of code is in fact Dart 2-clean?

**Comment by vsmenon** [comment]
@a-siva - note Nate's Dart 1 vs Dart 2 perf comment above.

**Comment by mit-mit** [comment]
This continues to cause friction when we work on making packages dart2 compatible. @a-siva is someone looking into the perf issues?

**Comment by dgrove** [comment]
We'll need to patch to make the SDK run in Dart 2 by default first.

**Comment by vsmenon** [comment]
Is that blocking? My understanding from @natebosch 's comment is that we can run from source to make progress on the rest - we'll need the snapshot for reasonable perf though.

**Comment by keertip** [comment]
We should start testing by running from source with --preview-dart-2 and in parallel look at dart2 snapshot generation so that we can have a dart 2 sdk build.

**Comment by nex3** [comment]
@vsmenon According to https://github.com/dart-lang/sdk/issues/32188#issuecomment-382179856, it looks like pub itself is in an okay state here, modulo the snapshot problem pointed out https://github.com/dart-lang/sdk/issues/32188#issuecomment-382526880.

Is there a way to detect at runtime if the current VM is running in Dart 2 mode?

**Comment by natebosch** [comment]
>  it looks like pub itself is in an okay state here

There is work we need to do in pub around invoking the right thing based on the Dart mode. If we're in Dart 2 we can't use the Dart 1 snapshot so we'll need to run from source (for now).

> Is there a way to detect at runtime if the current VM is running in Dart 2 mode?

`bool isDart2() => (<T>() => T)<String>() == String`

**Comment by nex3** [comment]
I'll see if I can take a crack at the snapshot issue on Monday.

**Comment by nex3** [comment]
When I disable running from snapshots in Dart 2, it looks like `Isolate.spawnUri()` isn't working properly. I've filed https://github.com/dart-lang/sdk/issues/32950 to track the SDK issue.

**Comment by dgrove** [comment]
#32950 has been fixed.

**Comment by kevmoo** [comment]
So trying to run `pub` directly fails – guessing it's because of the `pub snapshot`.

Tried this:

```console
$ dart --preview-dart-2 ~/source/github/pub/bin/pub.dart run test
Unable to spawn isolate: file:///Users/kevmoo/.pub-cache/hosted/pub.dartlang.org/front_end-0.1.0-alpha.9/lib/src/fasta/source/outline_builder.dart:736:49: Error: A value of type 'dart.core::List<dynamic>' can't be assigned to a variable of type 'dart.core::Iterable<fasta.formal_parameter_builder::FormalParameterBuilder<fasta.type_builder::TypeBuilder>>'.
Try changing the type of the left hand side, or casting the right hand side to 'dart.core::Iterable<fasta.formal_parameter_builder::FormalParameterBuilder<fasta.type_builder::TypeBuilder>>'.
        formals.setRange(count, formals.length, last);
                                                ^
```

**Comment by dgrove** [comment]
#33030 is closed. Is there any more work to do here, @natebosch ?

**Comment by kevmoo** [comment]
@dgrove – need to update `pub_semver` package – just told @natebosch

**Comment by kevmoo** [comment]
Darn. Still not fixed. I patched locally. Still getting

```console
$ DART_VM_OPTIONS=--preview-dart-2 pub
Missing root library
```

**Comment by natebosch** [comment]
@dgrove - there is much more to do.

- [ ] At a minimum we need to have special handling for this flag in the shell script in the SDK.
- [ ] To shave off ~10 seconds of the slowdown from fasta we need to generate a Dart 2 or kernel snapshot in the SDK.
- [ ] Ideally we need to completely overhaul the way snapshots are generated in pub so that we can generate app-jit snapshots on first run rather than static snapshots on `pub get`. In today's implementation we fall back on running from source for every run which is slow.

**Comment by natebosch** [comment]
> Still not fixed. I patched locally. Still getting

As you discovered, the reason is that we only ship the snapshot with the SDK. It looks like my 2nd bullet point from the message above is a requirement, not a nice-to-have.

**Comment by natebosch** [comment]
@a-siva you might need to sync past https://github.com/dart-lang/sdk/commit/4da831fd3ad72e4590cf007814d7b52c11d2a6d6

**Comment by kevmoo** [comment]
Pending CL - https://dart-review.googlesource.com/c/sdk/+/54743

**Comment by vsmenon** [comment]
@a-siva - @natebosch 's cl is breaking on the bots with:

```
C:/b/s/w/ir/cache/vpython/1bda7b/Scripts/python.exe ../../build/gn_run_binary.py compiled_action dart.exe --deterministic --packages=C:/b/s/w/ir/cache/builder/sdk/.packages --snapshot=gen/pub2.dart.snapshot --snapshot-depfile=C:/b/s/w/ir/cache/builder/sdk/out/ReleaseIA32/gen/pub2.dart.snapshot.d --preview-dart-2 --snapshot-kind=script C:/b/s/w/ir/cache/builder/sdk/third_party/pkg/pub/bin/pub.dart
Command failed: ./dart.exe --deterministic --packages=C:/b/s/w/ir/cache/builder/sdk/.packages --snapshot=gen/pub2.dart.snapshot --snapshot-depfile=C:/b/s/w/ir/cache/builder/sdk/out/ReleaseIA32/gen/pub2.dart.snapshot.d --preview-dart-2 --snapshot-kind=script C:/b/s/w/ir/cache/builder/sdk/third_party/pkg/pub/bin/pub.dart
output: c:/b/s/w/ir/cache/builder/sdk/.packages: Error: Unable to read '.packages' file:
  FileSystemException(uri=c:/b/s/w/ir/cache/builder/sdk/.packages; message=File c:/b/s/w/ir/cache/builder/sdk/.packages does not exist.).
```

I'm seeing similar with my DDC CL:
```
Command failed: ./dart --deterministic --packages=/Users/vsm/dart/sdk/.packages --snapshot=gen/dartdevc.dart.snapshot --snapshot-depfile=/Users/vsm/dart/sdk/xcodebuild/ReleaseX64/gen/dartdevc.dart.snapshot.d --preview-dart-2 --snapshot-kind=app-jit /Users/vsm/dart/sdk/pkg/dev_compiler/bin/dartdevc.dart --dart-sdk /Users/vsm/dart/sdk/sdk --dart-sdk-summary /Users/vsm/dart/sdk/xcodebuild/ReleaseX64/gen/utils/dartdevc/ddc_sdk.sum --library-root /Users/vsm/dart/sdk/pkg/dev_compiler -o dartdevc.js /Users/vsm/dart/sdk/pkg/dev_compiler/bin/dartdevc.dart
output: Error while initializing Kernel isolate
```

Any ideas?

**Comment by keertip** [comment]
@vsmenon, are you running from `xcodebuild/ReleaseX64/` directory?

**Comment by vsmenon** [comment]
Yes.

**Comment by a-siva** [comment]
This CL https://dart-review.googlesource.com/c/sdk/+/55589 fixes the .package not found error but now an assertion in the optimizing compiler is being triggered which I am investigating.

**Comment by alorenzen** [comment]
Spoke to @JekCharlsonYu, this was moved to Stable by mistake.

**Comment by a-siva** [comment]
https://dart-review.googlesource.com/c/sdk/+/55589 is being reviewed

**Comment by kevmoo** [comment]
Spoke too soon. Waiting on https://dart-review.googlesource.com/c/sdk/+/54743 from @natebosch

**Comment by natebosch** [comment]
@a-siva I'm still unable to build locally.

> Error while initializing Kernel isolate

Were you able to reproduce this error, did your CL fix it for you?

**Comment by natebosch** [comment]
the build error I'm seeing now might be flaky - it was for @vsmenon as well. I was able to get a successful build locally.

**Comment by kevmoo** [comment]
Hopefully fixed in https://github.com/dart-lang/sdk/commit/2ff2af79213b150f57799a69d2b7757fa71daa42

**Comment by natebosch** [comment]
We want https://github.com/dart-lang/pub/issues/1683 to fix the performance issues, but we're unblocked to run things (from source) now.
